bundler/lib/bundler/dep_proxy.rb

```diff
@@ -4,18 +4,27 @@ module Bundler
   class DepProxy
     attr_reader :__platform, :dep
 
+    @proxies = {}
+
+    def self.get_proxy(dep, platform)
+      @proxies[[dep, platform].hash] ||= new(dep, platform)
+    end
+
     def initialize(dep, platform)
       @dep = dep
       @__platform = platform
     end
 
-    def hash
-      @hash ||= [dep, __platform].hash
+    private_class_method :new
+    private :dup
+
+    # reset before calling
+    def self.setup!
+      @proxies = {}
```

Comment:
Is this in case bundler tries to resolve two different definitions in the same invokation? I'm not sure `bundler` whether ever does that... If it's for tests only, should we move this to test code?

Reply:
Agreed this is only for test code. There are other components that are setup in the resolver, so it seemed reasonable to do it there anyway.
I don't think we can remove this method, so it is about putting the call to `DepProxy.setup!` in a `before(:each)` block in `spec_helper.rb` if you think that would be better.

Reply:
I was thinking of something like `DepProxy.instance_variable_set(:@proxies, {})` inside tests, to make it clear that it's not a public method, but just some hack for testing several different resolutions in the same test run.

Reply:
Because all the tests run their own process, right? Let me look into it.

Separate question: I was wondering if there's a way to use pry inside a test run, or to do the test set up manually so that you can use the fake repos with `pry`. Is there any way to do that? I tried and failed miserably.

Reply:
Normally tests run on their own process, yeah, but I think resolver tests don't and that's  why I think we need to reset the dep_proxy cache between them so that they don't pollute each other.

Regarding `pry`, I normally use puts debugging when I need to debug tests that run subproceses (most), since we I tried using a REPL I failed miserably too.

Reply:
Need for `setup!` has been removed in 5aeaf9c8ee8dda2fa9fc812059e77d2d6d98502a.

Reply:
Awesome, nice catch! I'm thinking though that we should probably push this fix all the way down to `Gem::Requirement`, which is where the actual issue is, right?

Reply:
Maybe. It's intentional so the question is why is it that way? Canonical segments removes any trailing 0s on versions. So '2.0.0' becomes [2] and 2.1.0 becomes [2,1].

```ruby
  def hash # :nodoc:
    canonical_segments.hash
  end

  def canonical_segments
    @canonical_segments ||=
      _split_segments.map! do |segments|
        segments.reverse_each.drop_while {|s| s == 0 }.reverse
      end.reduce(&:concat)
  end
```

Reply:
But that's `Gem::Version`. I think for `Gem::Version` this is indeed intentional. I not 100% sure of the reason, but I think "2.0.0" and "2.0" are the same for most purposes, but it's useful to keep them different for others (for example, for display purposes?).

But for `Gem::Requirement` I don't think there's any reason why `~> 2.0.0` should ever be considered the same as `~> 2.0`.

Reply:
OK, I'm with you. let me have a go at it. However, that doesn't change the fact that we have to do it here, so I would do as a separate pull request. The reason is that this has to run against any version of rubygems, right?

Reply:
If we need a change in `rubygems` inside `bundler`, where normally backport the change into bundler when using older rubygems through a monkeypatch in `bundler/lib/bundler/rubygems_ext.rb`. That said, I wouldn't mind leaving it as it is for now and change it later.